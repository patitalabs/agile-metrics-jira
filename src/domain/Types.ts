import {
  PtsCollectorConfig,
  PtsMetricItem,
} from './project-tracking-system/Types';

export interface TeamMetricsRequest {
  shouldUpdateEntries: boolean;
  config: any;
}

export interface CoreMetricsService {
  publish(entries: any, shouldReplaceEntries: boolean);
}

export interface PtsCollectorService {
  fetch(ciCollectorConfig: PtsCollectorConfig): Promise<PtsMetricItem[]>;
}
