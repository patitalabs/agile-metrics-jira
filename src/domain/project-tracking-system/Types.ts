export interface PtsMetricItem {
  id: string;
  dataType: string;
  createdAt: Date;
  key: string;
  createdBy: string;
  issueType: string;
  movedForward: number;
  movedBackward: number;
  storyPoints: number;
  assignees: string[];
  tags: string[];
  created: Date;
  finished: Date;
  leadTime: number;
  cycleTime: number;
  commentCount: number;
  jiraProject: string;
  teamName: string;
  estimateHealth: number;
  rawEstimateHealth: number;
  numberOfBugs: number;
}

export class PtsCollectorConfig {
  teamId: number;
  teamName: string;
  workFlowType: 'sprint' | 'kanban';
  since: Date;
  until?: Date;
  workFlowMap?: { [name: string]: number };
  fields?: { [name: string]: string };
  estimateConfig?: { maxTime: number; estimationValues: number[] };

  constructor({
    teamId,
    since,
    until = null,
    workFlowMap,
    fields,
    teamName,
    workFlowType,
  }) {
    this.teamId = teamId;
    this.since = new Date(since);
    this.until = until ? new Date(until) : null;
    this.workFlowMap = workFlowMap;
    this.fields = fields;
    this.teamName = teamName;
    this.workFlowType = workFlowType ? workFlowType : 'sprint';
  }

  isKanban(): boolean {
    return this.workFlowType === 'kanban';
  }

  isSprint(): boolean {
    return this.workFlowType === 'sprint';
  }
}
