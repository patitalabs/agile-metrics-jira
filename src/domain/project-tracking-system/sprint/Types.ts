export interface TaskStatistics {
  moveBackward: number;
  moveForward: number;
  movedToDev: Date;
}
