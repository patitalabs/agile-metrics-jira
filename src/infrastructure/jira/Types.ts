export class Sprint {
  id: number;
  name: string;
  isoStartDate: Date;
  isoEndDate: Date;
}

export class Task {
  key: string;
  typeName: string;
  statusName: string;
  assignee: string;
  epic: string;
  sprint?: Sprint;
  created: Date;
  projectName: string;
  teamName: string;
  createdBy: string;
  resolutionDate: Date;
  storyPoints: number;
  labels: string[];
  subtasks: Subtask[];
  histories: { [type: string]: HistoryEntry[] };
  numberOfComments: number;
  numberOfBugs: number;
}

export class Subtask {
  key: string;
  statusName: string;
}

export class HistoryItem {
  field: string;
  fromString: string;
  toString: string;
}

export class HistoryEntry {
  category: string;
  created: Date;
  items: HistoryItem[];
}

export interface JiraConfig {
  teamId: number;
  teamName: string;
  since: Date;
  until?: Date;
  workFlowMap?: { [name: string]: number };
  workFlowType: 'sprint' | 'kanban';
  fields?: { [name: string]: string };
}

export interface IssueDetails {
  subtasks: Subtask[];
  histories: { [type: string]: HistoryEntry[] };
  labels: string[];
  created: Date;
  createdBy: string;
  resolutionDate: Date;
  projectName: string;
  teamName: string;
  numberOfComments: number;
  numberOfBugs: number;
}

export interface JiraService {
  completedSprintsSince(
    teamId: number,
    since: Date,
    until: Date
  ): Promise<Sprint[]>;

  sprintData(jiraConfig: JiraConfig, sprint: Sprint): Promise<Task[]>;

  completedKanbanIssuesSince(jiraConfig: JiraConfig): Promise<Task[]>;
}

export interface JiraClient {
  getData(url: string): Promise<any>;
}

export interface JiraRepository {
  issueDetailsWithChangelog(
    jiraConfig: JiraConfig,
    issueId: string
  ): Promise<IssueDetails>;
  sprintData(jiraConfig: JiraConfig, sprint: Sprint): Promise<Task[]>;

  completedSprints(
    teamId: number,
    since: Date,
    until?: Date
  ): Promise<Sprint[]>;

  completedKanbanIssuesSince(jiraConfig: JiraConfig): Promise<Task[]>;
}
