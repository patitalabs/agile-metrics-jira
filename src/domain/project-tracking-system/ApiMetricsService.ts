import { Logger } from '../../config/Logger';
import {
  PtsCollectorService,
  CoreMetricsService,
  TeamMetricsRequest,
} from '../Types';

export class ApiMetricsService {
  constructor(
    private readonly coreMetricsService: CoreMetricsService,
    private readonly ciCollectorService: PtsCollectorService
  ) {}

  public async metricsForRequest(
    teamMetricsRequest: TeamMetricsRequest
  ): Promise<void> {
    try {
      Logger.info(`teamMetricsRequest: ${JSON.stringify(teamMetricsRequest)}`);
      const configurationDescriptors =
        ApiMetricsService.createConfigurationDescriptorsForRequest(
          teamMetricsRequest
        );
      Logger.info(
        `created configurations: ${JSON.stringify(configurationDescriptors)}`
      );

      for (const configurationDescriptor of configurationDescriptors) {
        const scmMetricItems = await this.ciCollectorService.fetch(
          configurationDescriptor.config
        );
        await this.coreMetricsService.publish(
          scmMetricItems,
          configurationDescriptor.shouldUpdateEntries
        );
      }
      Logger.info('Done!');
    } catch (error) {
      Logger.error(error);
    }
  }

  static createConfigurationDescriptorsForRequest(
    teamMetricsRequest: TeamMetricsRequest
  ): any[] {
    if (teamMetricsRequest.config) {
      return [{ ...teamMetricsRequest }];
    }

    throw Error(
      `Unable to create configuration for request ${JSON.stringify(
        teamMetricsRequest
      )}`
    );
  }
}
