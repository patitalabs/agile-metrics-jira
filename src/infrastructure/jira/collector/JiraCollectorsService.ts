import { JiraConfig, JiraService, Sprint, Task } from '../Types';
import { JiraMetricConverter } from './JiraMetricConverter';
import { PtsCollectorService } from '../../../domain/Types';
import {
  PtsCollectorConfig,
  PtsMetricItem,
} from '../../../domain/project-tracking-system/Types';
import { Utils } from '../../../utils/Utils';

export class JiraCollectorsService implements PtsCollectorService {
  constructor(private readonly jiraService: JiraService) {}

  public async fetch(
    ptsCollectorConfig: PtsCollectorConfig
  ): Promise<PtsMetricItem[]> {
    const tasks: Task[] = await this.tasks(ptsCollectorConfig);
    return tasks
      .filter((task) => task != null)
      .map((task) =>
        JiraMetricConverter.toMetricItem(ptsCollectorConfig, task)
      );
  }

  private async tasks(ptsCollectorConfig: PtsCollectorConfig): Promise<Task[]> {
    let tasks = [];
    if (ptsCollectorConfig.isKanban()) {
      tasks = await this.tasksForKanban(ptsCollectorConfig);
    } else {
      tasks = await this.tasksForSprint(ptsCollectorConfig);
    }
    return tasks;
  }

  private async tasksForSprint(
    ptsCollectorConfig: PtsCollectorConfig
  ): Promise<Task[]> {
    const sprints: Sprint[] = await this.jiraService.completedSprintsSince(
      ptsCollectorConfig.teamId,
      ptsCollectorConfig.since,
      ptsCollectorConfig.until
    );

    const taskPromises: Promise<Task[]>[] = sprints.map((sprint) =>
      this.sprintDetails(ptsCollectorConfig, sprint)
    );

    return Utils.flatMap((item) => item, await Promise.all(taskPromises));
  }

  private async tasksForKanban(
    ptsCollectorConfig: PtsCollectorConfig
  ): Promise<Task[]> {
    const jiraConfig = JiraCollectorsService.toJiraConfig(ptsCollectorConfig);
    return this.jiraService.completedKanbanIssuesSince(jiraConfig);
  }

  private static toJiraConfig(
    ptsCollectorConfig: PtsCollectorConfig
  ): JiraConfig {
    return { ...ptsCollectorConfig };
  }

  private async sprintDetails(
    ptsCollectorConfig: PtsCollectorConfig,
    sprint: Sprint
  ): Promise<Task[]> {
    const jiraConfig = JiraCollectorsService.toJiraConfig(ptsCollectorConfig);
    return this.jiraService.sprintData(jiraConfig, sprint);
  }
}
