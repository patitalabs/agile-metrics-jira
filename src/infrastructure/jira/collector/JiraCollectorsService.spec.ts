import {
  HistoryEntry,
  JiraConfig,
  JiraService,
  Sprint,
  Subtask,
  Task,
} from '../Types';
import { JiraCollectorsService } from './JiraCollectorsService';
import { PtsCollectorConfig } from '../../../domain/project-tracking-system/Types';

function testSubtasks(): Subtask[] {
  return [{ key: 'someKey-12', statusName: 'Done' }];
}

function testHistories(): { [type: string]: HistoryEntry[] } {
  const historyEntry: HistoryEntry = {
    category: 'string',
    created: new Date('2018-12-04'),
    items: [
      {
        field: 'status',
        fromString: 'Open',
        toString: 'Done',
      },
    ],
  };
  return {
    status: [historyEntry],
  };
}

function testTask(): Task {
  return {
    key: 'someKey-1',
    typeName: 'Story',
    statusName: 'Done',
    assignee: 'someone',
    epic: 'someEpic',
    sprint: testSprint(),
    created: new Date('2018-12-02'),
    projectName: 'someProject',
    teamName: 'someTeam',
    createdBy: 'someone',
    resolutionDate: new Date('2018-12-06'),
    storyPoints: 2,
    labels: ['label1', 'label2'],
    subtasks: testSubtasks(),
    histories: testHistories(),
    numberOfComments: 4,
    numberOfBugs: 1,
  };
}

function testSprint(): Sprint {
  return {
    id: 45,
    name: 'sprint name',
    isoStartDate: new Date('2018-12-03'),
    isoEndDate: new Date('2018-12-11'),
  };
}

describe('JiraCollectorsService', () => {
  const jiraService: JiraService = {
    completedSprintsSince: (
      teamId: number,
      referenceDate: Date
    ): Promise<Sprint[]> => Promise.resolve([testSprint()]),

    sprintData: (jiraConfig: JiraConfig, sprint: Sprint): Promise<Task[]> =>
      Promise.resolve([testTask()]),

    completedKanbanIssuesSince: (jiraConfig: JiraConfig): Promise<Task[]> =>
      Promise.resolve([testTask()]),
  };

  const jiraCollectorsService: JiraCollectorsService =
    new JiraCollectorsService(jiraService);

  const config = ({ workFlowType }): PtsCollectorConfig =>
    new PtsCollectorConfig({
      teamId: 68,
      teamName: 'someTeamName',
      since: '2018-11-20',
      workFlowMap: null,
      workFlowType,
      fields: null,
      until: '2020-11-20',
    });

  it('should fetch jiraMetrics for sprint', async () => {
    const collectorConfig = config({ workFlowType: 'sprint' });

    const data = await jiraCollectorsService.fetch(collectorConfig);
    expect(data).toMatchSnapshot();
  });

  it('should fetch jiraMetrics for kanban', async () => {
    const collectorConfig = config({ workFlowType: 'kanban' });

    const data = await jiraCollectorsService.fetch(collectorConfig);
    expect(data).toMatchSnapshot();
  });
});
