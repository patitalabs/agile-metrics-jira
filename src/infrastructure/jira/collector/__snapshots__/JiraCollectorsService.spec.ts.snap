// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JiraCollectorsService should fetch jiraMetrics for kanban 1`] = `
Array [
  Object {
    "assignees": Array [
      "someone",
    ],
    "commentCount": 4,
    "created": 2018-12-02T00:00:00.000Z,
    "createdAt": 2018-12-06T00:00:00.000Z,
    "createdBy": "someone",
    "cycleTime": 2,
    "dataType": "PTS",
    "estimateHealth": 0,
    "finished": 2018-12-06T00:00:00.000Z,
    "id": "b8c1af4128307d9bac209e2f9504b3d1992cda17a1e4a6144f43ad5fe2bb52da000ca14e9322b17fe105d72d809214d0a86cfaff844b2e8bcff2e14d0b7b665e",
    "issueType": "Story",
    "jiraProject": "someProject",
    "key": "someKey-1",
    "leadTime": 4,
    "movedBackward": 0,
    "movedForward": 1,
    "numberOfBugs": 1,
    "rawEstimateHealth": 0.00001,
    "storyPoints": 2,
    "tags": Array [
      "label1",
      "label2",
    ],
    "teamName": "someTeam",
  },
]
`;

exports[`JiraCollectorsService should fetch jiraMetrics for sprint 1`] = `
Array [
  Object {
    "assignees": Array [
      "someone",
    ],
    "commentCount": 4,
    "created": 2018-12-02T00:00:00.000Z,
    "createdAt": 2018-12-06T00:00:00.000Z,
    "createdBy": "someone",
    "cycleTime": 2,
    "dataType": "PTS",
    "estimateHealth": 0,
    "finished": 2018-12-06T00:00:00.000Z,
    "id": "b8c1af4128307d9bac209e2f9504b3d1992cda17a1e4a6144f43ad5fe2bb52da000ca14e9322b17fe105d72d809214d0a86cfaff844b2e8bcff2e14d0b7b665e",
    "issueType": "Story",
    "jiraProject": "someProject",
    "key": "someKey-1",
    "leadTime": 4,
    "movedBackward": 0,
    "movedForward": 1,
    "numberOfBugs": 1,
    "rawEstimateHealth": 0.00001,
    "storyPoints": 2,
    "tags": Array [
      "label1",
      "label2",
    ],
    "teamName": "someTeam",
  },
]
`;
